File: clients/macos/SynthesisApp/Sources/SynthesisMCP/MCPClient.swift
import Foundation

public final class MCPClient: ObservableObject {
    public let baseURL: URL
    private let mcpURL: URL
    private let session: URLSession
    private var nextId = 0
    private var sessionId: String?

    @Published public var initialized = false

    public init(baseURL: URL, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.mcpURL = baseURL.appendingPathComponent("mcp")
        self.session = session
    }

    // MARK: - JSON-RPC transport

    private func nextRequestId() -> Int {
        nextId += 1
        return nextId
    }

    private func send(method: String, params: [String: Any]? = nil) async throws -> JSONRPCResponse {
        var req = URLRequest(url: mcpURL)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json, text/event-stream", forHTTPHeaderField: "Accept")
        if let sid = sessionId {
            req.setValue(sid, forHTTPHeaderField: "Mcp-Session-Id")
        }

        let rpcRequest = JSONRPCRequest(
            id: nextRequestId(),
            method: method,
            params: params.map { AnyCodable($0) }
        )
        req.httpBody = try JSONEncoder().encode(rpcRequest)

        let (data, response) = try await session.data(for: req)

        if let httpResponse = response as? HTTPURLResponse,
           let sid = httpResponse.value(forHTTPHeaderField: "Mcp-Session-Id") {
            sessionId = sid
        }

        let contentType = (response as? HTTPURLResponse)?
            .value(forHTTPHeaderField: "Content-Type") ?? ""

        if contentType.contains("text/event-stream") {
            let jsonData = try extractSSEData(from: data)
            return try JSONDecoder().decode(JSONRPCResponse.self, from: jsonData)
        }
        return try JSONDecoder().decode(JSONRPCResponse.self, from: data)
    }

    private func sendNotification(method: String, params: [String: Any]? = nil) async throws {
        var req = URLRequest(url: mcpURL)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json, text/event-stream", forHTTPHeaderField: "Accept")
        if let sid = sessionId {
            req.setValue(sid, forHTTPHeaderField: "Mcp-Session-Id")
        }

        // Notifications have no id
        var body: [String: Any] = ["jsonrpc": "2.0", "method": method]
        if let p = params { body["params"] = p }
        req.httpBody = try JSONSerialization.data(withJSONObject: body)

        _ = try await session.data(for: req)
    }

    // MARK: - MCP Lifecycle

    public func initialize() async throws {
        let resp = try await send(method: "initialize", params: [
            "protocolVersion": "2025-03-26",
            "capabilities": [:] as [String: Any],
            "clientInfo": ["name": "SynthesisApp", "version": "0.0.1"]
        ])

        if resp.error != nil {
            throw MCPError(message: resp.error?.message ?? "Initialize failed")
        }

        // Send initialized notification
        try await sendNotification(method: "notifications/initialized")

        await MainActor.run { initialized = true }
    }

    // MARK: - Tools

    public func listTools() async throws -> [MCPTool] {
        let resp = try await send(method: "tools/list")

        guard let result = resp.result?.value as? [String: Any],
              let tools = result["tools"] as? [[String: Any]] else {
            return []
        }

        return tools.compactMap { t in
            guard let name = t["name"] as? String,
                  let desc = t["description"] as? String else { return nil }
            return MCPTool(name: name, description: desc)
        }
    }

    public func callTool(name: String, arguments: [String: Any]) async throws -> [String: Any]? {
        let resp = try await send(method: "tools/call", params: [
            "name": name,
            "arguments": arguments
        ])

        if let err = resp.error {
            throw MCPError(message: err.message)
        }

        // Parse MCP content array: [{type:"text", text:"..."}]
        guard let result = resp.result?.value as? [String: Any],
              let content = result["content"] as? [[String: Any]],
              let first = content.first,
              let text = first["text"] as? String,
              let data = text.data(using: .utf8) else {
            return nil
        }

        return try JSONSerialization.jsonObject(with: data) as? [String: Any]
    }

    // MARK: - SSE Parsing

    private func extractSSEData(from data: Data) throws -> Data {
        guard let text = String(data: data, encoding: .utf8) else {
            throw MCPError(message: "Invalid UTF-8 in SSE response")
        }
        for line in text.components(separatedBy: "\n") {
            if line.hasPrefix("data: "), line.count > 6 {
                let json = String(line.dropFirst(6))
                if let d = json.data(using: .utf8) { return d }
            }
            if line.hasPrefix("data:"), !line.hasPrefix("data: ") {
                let json = String(line.dropFirst(5))
                if !json.isEmpty, let d = json.data(using: .utf8) { return d }
            }
        }
        throw MCPError(message: "No data field found in SSE response")
    }
}

public struct MCPError: LocalizedError {
    public let message: String
    public var errorDescription: String? { message }

    public init(message: String) {
        self.message = message
    }
}

File: clients/lovable-web/src/lib/mcpClient.ts
type JsonObject = Record<string, unknown>;

type JsonRpcResponse = {
  jsonrpc: "2.0";
  id?: number;
  result?: unknown;
  error?: { code: number; message: string; data?: unknown };
};

export class MCPClient {
  private readonly endpoint: string;
  private sessionId?: string;
  private requestId = 0;
  private isInitialized = false;

  constructor(baseUrl: string) {
    this.endpoint = new URL("/mcp", baseUrl).toString();
  }

  async initialize(): Promise<void> {
    if (this.isInitialized) return;

    await this.send("initialize", {
      protocolVersion: "2025-03-26",
      capabilities: {},
      clientInfo: {
        name: "Synthesis Lovable Web",
        version: "0.0.1"
      }
    });

    await this.send("notifications/initialized", undefined, true);
    this.isInitialized = true;
  }

  async callTool<T>(name: string, args: JsonObject): Promise<T> {
    const result = (await this.send("tools/call", {
      name,
      arguments: args
    })) as { content?: Array<{ type?: string; text?: string }> } | undefined;

    const text = result?.content?.find((item) => item.type === "text")?.text;
    if (!text) {
      throw new Error(`Tool ${name} returned no JSON text payload`);
    }

    try {
      return JSON.parse(text) as T;
    } catch {
      throw new Error(`Tool ${name} returned invalid JSON text payload`);
    }
  }

  private nextRequestId(): number {
    this.requestId += 1;
    return this.requestId;
  }

  private parseSSEPayload(raw: string): string {
    for (const line of raw.split("\n")) {
      if (!line.startsWith("data:")) continue;
      const payload = line.slice(5).trimStart();
      if (payload.length > 0) return payload;
    }
    throw new Error("SSE response did not include a data payload");
  }

  private async send(method: string, params?: JsonObject, notification = false): Promise<unknown> {
    const body: JsonObject = {
      jsonrpc: "2.0",
      method
    };

    if (!notification) {
      body.id = this.nextRequestId();
    }
    if (params) {
      body.params = params;
    }

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      Accept: "application/json, text/event-stream"
    };
    if (this.sessionId) {
      headers["Mcp-Session-Id"] = this.sessionId;
    }

    const response = await fetch(this.endpoint, {
      method: "POST",
      headers,
      body: JSON.stringify(body)
    });

    const nextSessionId = response.headers.get("Mcp-Session-Id");
    if (nextSessionId) {
      this.sessionId = nextSessionId;
    }

    const responseText = await response.text();

    if (!response.ok) {
      throw new Error(responseText || `HTTP ${response.status}`);
    }

    if (notification) {
      return undefined;
    }

    const contentType = response.headers.get("Content-Type") ?? "";
    const jsonText = contentType.includes("text/event-stream")
      ? this.parseSSEPayload(responseText)
      : responseText;

    let rpc: JsonRpcResponse;
    try {
      rpc = JSON.parse(jsonText) as JsonRpcResponse;
    } catch {
      throw new Error("MCP response is not valid JSON");
    }

    if (rpc.error) {
      throw new Error(rpc.error.message);
    }

    return rpc.result;
  }
}
